src/odemis/acq/milling/millmng.py

```diff
@@ -0,0 +1,225 @@
+
+        # error management
+        exception = False
+        try:
+
+            # acquire a reference image at the imaging settings
+            if align_at_milling_current:
+                self._future.running_subf = acquire([self.fib_stream])
+                data, _ = self._future.running_subf.result()
+                ref_image = data[0]
+
+            # set the milling state
+            self.fibsem.clear_patterns()
+            self.fibsem.set_default_patterning_beam_type(milling_channel)
+            self.fibsem.set_high_voltage(milling_voltage, milling_channel)
+            self.fibsem.set_beam_current(milling_current, milling_channel)
+            # self.fibsem.set_field_of_view(milling_fov, milling_channel) # tmp: disable until matched in gui
+            self.fibsem.set_patterning_mode(milling_mode)
+
+            # acquire a new image at the milling settings and align
+            if align_at_milling_current:
+                self._future.running_subf = acquire([self.fib_stream])
+                data, _ = self._future.running_subf.result()
+                new_image = data[0]
+                align_reference_image(ref_image, new_image, self.ion_beam)
+
+            # draw milling patterns to microscope
+            for pattern in settings.generate():
+                if isinstance(pattern, RectanglePatternParameters):
+                    self.fibsem.create_rectangle(pattern.to_dict())
+                else:
+                    raise NotImplementedError(f"Pattern {pattern} not supported") # TODO: support other patterns
+
+            # estimate the milling time
+            estimated_time = self.fibsem.estimate_milling_time()
+            self._future.set_end_time(time.time() + estimated_time)
+
+            # start patterning (async)
+            self.fibsem.start_milling()
+
+            # wait for milling to finish
+            elapsed_time = 0
+            wait_time = 5
+            while self.fibsem.get_patterning_state() == "Running":
+
+                with self._future._task_lock:
+                    if self._future.cancelled() == CANCELLED:
+                        raise CancelledError()
+
+                logging.debug(f"Milling in progress... elapsed time: {elapsed_time} s, estimated time:  {estimated_time} s")
+                time.sleep(wait_time)
+                elapsed_time += wait_time
+
+        except CancelledError as ce:
+            logging.debug(f"Cancelled milling: {ce}")
+            raise
+        except Exception as e:
+            logging.exception(f"Error while milling: {e}")
+            raise
+        finally:
+            # restore imaging state
+            self.fibsem.set_beam_current(imaging_current, milling_channel)
+            self.fibsem.set_high_voltage(imaging_voltage, milling_channel)
+            self.fibsem.set_field_of_view(imaging_fov, milling_channel)
+            #self.fibsem.set_channel(milling_channel) # TODO: expose on server
+            self.fibsem.set_active_view(2)
```

Comment:
Can you add a short comment to explain why you need to set the active view? And a related question, if you should set the active view before clearing the patterns, shouldn't you also do that on line 120?

Reply:
it is to ensure the fib view is selected, on 120 probably should as well. instead i've added it to the client for clear_patterns because at the moment patterns are only drawn on fib (active_view = 2). SO i've removed the calls here.